lxd/storage/drivers/driver_ceph_volumes.go

```diff
@@ -644,7 +644,192 @@ func (d *ceph) CreateVolumeFromMigration(vol VolumeCopy, conn io.ReadWriteCloser
+	if lastCommonSnapshotIndex >= 0 {
+		_, lastCommonSnapshotName, _ = api.GetParentAndSnapshotName(vol.Snapshots[lastCommonSnapshotIndex].name)
+	}
+
+	if lastCommonSnapshotName != "" {
+		// Remove all snapshots from the target that will get refreshed.
+		// Those are all the snapshots after the last common snapshot.
+		lastCommonSnapshotFound := false
+		for _, targetSnapshot := range vol.Snapshots {
+			_, targetSnapshotName, _ := api.GetParentAndSnapshotName(targetSnapshot.name)
+			if targetSnapshotName == lastCommonSnapshotName {
+				// The last common snapshot was found.
+				// Continue the loop and start to delete all of the following snapshots.
+				lastCommonSnapshotFound = true
+				continue
+			}
+
+			// Delete all of the snapshots after the last common snapshot.
+			if lastCommonSnapshotFound {
+				ok, err := d.hasVolume(d.getRBDVolumeName(vol.Volume, fmt.Sprintf("snapshot_%s", targetSnapshotName), false, false))
+				if err != nil {
+					return err
+				}
+
+				// The snapshot does not exist on the target.
+				if !ok {
+					continue
+				}
+
+				// Delete the snapshot if its order is out of sync.
+				// This happens if not the latest snapshot on the target side gets deleted and requires refresh.
+				// The VMs filesystem volume snapshot will not be deleted.
+				// It already got refreshed using the generic approach.
+				_, err = d.deleteVolumeSnapshot(vol.Volume, fmt.Sprintf("snapshot_%s", targetSnapshotName))
+				if err != nil {
+					return err
+				}
+			}
+		}
+
+		// Before syncing the snapshots restore the target to the last common snapshot.
+		// This is required so that the target volume is at the exact same state as the source volume.
+		// We can then use the Ceph RBD export-diff/import-diff functions to create the delta
+		// between the latest snapshot and source volume and apply it on the target volume.
+		// The VMs filesystem volume will not be restored.
+		// It already got refreshed using the generic approach.
+		err := d.restoreVolume(vol.Volume, vol.Snapshots[lastCommonSnapshotIndex], op)
+		if err != nil {
+			return err
+		}
+	} else {
+		// There isn't a common snapshot on the target volume.
+		// Delete the volume as we will create a new sparse copy.
+		// The VMs filesystem volume will not be deleted.
+		// It already got refreshed using the generic approach.
+		_, err := d.deleteVolume(vol.Volume)
+		if err != nil {
+			return err
+		}
+
+		// Recreate the volume.
+		// A filler is not required since the source diff will be applied.
+		// The snapshots will get synced at a later step.
+		// If the volume is of type snapshot, it will get recreated later from copy.
+		if !srcVol.IsSnapshot() {
+			// The VMs filesystem volume will not be recreated.
+			// It already got refreshed using the generic approach.
+			err := d.rbdCreateVolume(vol.Volume, vol.ConfigSize())
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	if srcVol.IsSnapshot() {
+		// The target volume was just deleted.
```

Comment:
where does this happen?

Reply:
Maybe another comment would be supportive here.
In case a snapshot gets refreshed, it doesn't have any snapshots itself. Therefore the target volume get deleted first (see the lines above where it runs `_, err := d.deleteVolume(vol.Volume)`) so that the snapshot can be recreated from copy.

Reply:
yeah that would be useful thanks.